transformer_engine/common/permutation/permutation.cu

```diff
@@ -0,0 +1,296 @@
+      }
+    }
+  }
+}
+
+template <typename T, bool FWD>
+void moe_permutation_launcher(const void *input_, void *output_, const int *sorted_row_id,
+                              int *row_id_map, const float *prob, const int num_rows,
+                              const int num_topK, const int num_cols, const int num_out_tokens,
+                              cudaStream_t stream, float *prob_grad, const void *input_fwd_) {
+  using TCompute = typename std::conditional<(std::is_same<T, __nv_fp8_e5m2>::value ||
+                                              std::is_same<T, __nv_fp8_e4m3>::value),
+                                             half, T>::type;
+
+  static constexpr int kElementsPerAccess = 16 / sizeof(T);
+
+  const T *input = reinterpret_cast<const T *>(input_);
+  T *output = reinterpret_cast<T *>(output_);
+  const T *input_fwd = reinterpret_cast<const T *>(input_fwd_);
+
+  if (FWD) {
+    if (prob == nullptr) {
+      if (input_fwd == nullptr) {
+        // Permute fwd
+        int threads = 64;
+        int blocks = (num_rows * num_topK + threads - 1) / threads;
+        moe_permute_row_map<<<blocks, threads, 0, stream>>>(sorted_row_id, row_id_map, num_rows,
+                                                            num_topK, num_out_tokens);
+
+        blocks = num_rows;
+        threads = std::min(num_cols / kElementsPerAccess, 1024);
+        moe_permute_kernel<T, TCompute, 128, false><<<blocks, threads, 0, stream>>>(
+            input, nullptr, output, nullptr, nullptr, row_id_map, num_rows, num_topK, num_cols);
+      } else {
+        // Unpermute bwd without probs for topK == 1
+        int blocks = num_rows;
+        int threads = 32;
+
+        moe_permute_kernel<T, TCompute, 1, false><<<blocks, threads, 0, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, num_topK, num_cols);
+      }
+    } else {
+      // Unpermute bwd with probs
+      int blocks = num_rows;
+      int threads = 32;
+      size_t smem_bytes = num_topK * sizeof(TCompute);
+
+      if (num_topK <= 8) {
+        moe_permute_kernel<T, TCompute, 8, true><<<blocks, threads, smem_bytes, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, num_topK, num_cols);
+      } else if (num_topK <= 16) {
+        moe_permute_kernel<T, TCompute, 16, true><<<blocks, threads, smem_bytes, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, num_topK, num_cols);
+      } else if (num_topK <= 32) {
+        moe_permute_kernel<T, TCompute, 32, true><<<blocks, threads, smem_bytes, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, num_topK, num_cols);
+      } else if (num_topK <= 64) {
+        moe_permute_kernel<T, TCompute, 64, true><<<blocks, threads, smem_bytes, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, num_topK, num_cols);
+      } else if (num_topK <= 128) {
```

Comment:
Each thread stores an array of 128 floats on the register memory. Will this cause low GPU occupancy, thus the performance overall?

Reply:
In actual use, the value of K is generally small, and the chances of K >= 64 are very rare.
Actually, `moe_permute_kernel` is used for two paths, permute fwd and unpermute bwd. The register array is only for the case of unpermute bwd with probs, and in this case it's a fused kernel which can calculate two gradiants (act and probs) in a single kernel. The problem is that the calculation of act.grad is along the hidden dim, while whereas the calculation of probs.grad is along the topK dim. In order to fuse these two calculations in a single kernel, an array is needed.